File: src/parser_utils/__tests__/flow_polymorphic_ast_mapper_test.ml
(**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

open OUnit2

class ['a] mapper =
  object
    inherit [Loc.t, 'a, Loc.t, 'a] Flow_polymorphic_ast_mapper.mapper

    method on_loc_annot (x : Loc.t) = x

    method on_type_annot (x : 'a) = x
  end

(* these tests don't do much other than check that the mapper doesn't raise
    exceptions *)

let run_mapper source =
  let (ast, _) = Parser_flow.program source in
  let mapper = new mapper in
  let _ = mapper#program ast in
  ()

let tests =
  "polymorphic ast mapper"
  >::: [ ( "simple"
         >:: fun _ ->
         let source = "function foo() { (5 * 3); 4; (6 + 4); }" in
         run_mapper source ) ]

File: rust_port/crates/flow_parser/src/flow_polymorphic_ast_mapper_test.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use dupe::Dupe;

use crate::loc::Loc;
use crate::parse_program_without_file;
use crate::polymorphic_ast_mapper;
use crate::polymorphic_ast_mapper::LocMapper;

struct Mapper;

impl LocMapper<Loc, Loc, Loc, Loc> for Mapper {
    fn on_loc_annot(&mut self, x: &Loc) -> Result<Loc, !> {
        Ok(x.dupe())
    }

    fn on_type_annot(&mut self, x: &Loc) -> Result<Loc, !> {
        Ok(x.dupe())
    }
}

// these tests don't do much other than check that the mapper doesn't raise exceptions
fn run_mapper(source: &str) {
    let (ast, _errors) = parse_program_without_file(true, None, None, Ok(source));
    let mut mapper = Mapper;
    let Ok(_mapped) = polymorphic_ast_mapper::program(&mut mapper, &ast);
}

#[test]
fn simple() {
    let source = "function foo() { (5 * 3); 4; (6 + 4); }";
    run_mapper(source);
}
